python/ray/serve/tests/test_standalone2.py

```diff
@@ -1175,6 +1179,72 @@ def test_deploy_one_app_failed(self, client: ServeControllerClient):
             == ApplicationStatus.DEPLOY_FAILED
         )
 
+    def test_deploy_with_route_prefix_conflict(self, client: ServeControllerClient):
+        world_import_path = "ray.serve.tests.test_config_files.world.DagNode"
+        pizza_import_path = "ray.serve.tests.test_config_files.pizza.serve_dag"
+        test_config = {
+            "host": "127.0.0.1",
+            "port": 8000,
+            "applications": [
+                {
+                    "name": "app1",
+                    "route_prefix": "/app1",
+                    "import_path": world_import_path,
+                },
+                {
+                    "name": "app2",
+                    "route_prefix": "/app2",
+                    "import_path": pizza_import_path,
+                },
+            ],
+        }
+
+        client.deploy_apps(ServeDeploySchema(**test_config))
+
+        wait_for_condition(
+            lambda: requests.get("http://localhost:8000/app1").text == "wonderful world"
+        )
+        wait_for_condition(
+            lambda: requests.post("http://localhost:8000/app2", json=["ADD", 2]).json()
+            == "4 pizzas please!"
+        )
+
+        # Buffer time
+        time.sleep(1)
+
+        test_config["applications"][1] = {
+            "name": "app3",
+            "route_prefix": "/app2",
+            "import_path": world_import_path,
+        }
+
+        client.deploy_apps(ServeDeploySchema(**test_config))
+
+        def check():
+            serve_details = ServeInstanceDetails(
+                **ray.get(client._controller.get_serve_instance_details.remote())
+            )
+            app1_running = (
+                "app1" in serve_details.applications
+                and serve_details.applications["app1"].status == "RUNNING"
+            )
+            app3_running = (
+                "app3" in serve_details.applications
+                and serve_details.applications["app3"].status == "RUNNING"
+            )
+            app2_gone = "app2" not in serve_details.applications
```

Comment:
can we make app2 is under deleting status in this test case?

Reply:
IIUC, the test case doesn't guarantee the code added in this pr will be tested.

Reply:
The goal of this PR is to make sure a route prefix conflict between two versions of configs will not cause issues.

There is no way to guarantee that one app will be deleting when the new one is deploying (it will probably causing flaky test failures). If we want to check the details of how application state changes with time, I would need to add an unit test with deployments mocked out, but I wasn't sure if we should enforce "when one app is deleting, the other one should be able to deploy" since this is a temporary measure. However, I'm open to adding it.

Reply:
Synced offline - I've added a test to `test_application_state`!